### PR TITLE
MR: apply row-level delete files when reading

### PR DESCRIPTION
**Pull request by chenjunjiedada**
This applies row-level delete files when reading for IcebergInputFormat. This also includes changes from https://github.com/apache/iceberg/pull/985.

This also refactors the deletes read unit tests to a separated base test class to avoid duplication.

### CODE REVIEW

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java [hunk 1]
```diff
@@ -148,15 +144,15 @@ public void testMixedPositionAndEqualityDeletes() throws IOException {
     );
 
     DeleteFile eqDeletes = FileHelpers.writeDeleteFile(
-        table, Files.localOutput(temp.newFile()), Row.of(0), dataDeletes, dataSchema);
+        table, Files.localOutput(temp.newFile()), TestHelpers.Row.of(0), dataDeletes, dataSchema);
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java:147]: Can you import this class directly to avoid so many changes in this file?

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java [hunk 2]
```diff
@@ -269,4 +265,5 @@ private StructLikeSet rowSetWithoutIds(int... idsToRemove) {
         .forEach(set::add);
     return set;
   }
+
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java:268]: Nit: unnecessary whitespace change

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java [hunk 3]
```diff
@@ -129,7 +133,7 @@
           // TODO: We do not support residual evaluation for HIVE and PIG in memory data model yet
           checkResiduals(task);
         }
-        splits.add(new IcebergSplit(conf, task));
+        splits.add(new IcebergSplit(conf, task, table.io(), table.encryption()));
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:136]: While I would like to get the encryption manager and io changes in, I don't think that they should be mixed into this commit. Was it necessary to do this for some reason?

File: data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import org.junit.Before;
+
+import static org.apache.iceberg.types.Types.NestedField.required;
+
+public class GenericReaderDeletesTest extends DeletesReadTest {
+  // Schema passed to create tables
+  public static final Schema SCHEMA = new Schema(
+      required(1, "id", Types.IntegerType.get()),
+      required(2, "data", Types.StringType.get())
+  );
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java:38]: Why not put the schema and spec in the parent class, `DeletesReadTest`? The data it generates is for this schema.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java [hunk 5]
```diff
@@ -92,6 +72,22 @@ public void testEqualityDeletes() throws IOException {
     Assert.assertEquals("Table should contain expected rows", expected, actual);
   }
 
+  protected void generateTestData() throws IOException {
+    this.records = new ArrayList<>();
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java:76]: We prefer using `Lists.newArrayList()`

File: data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  // Partition spec used to create tables
+  static final PartitionSpec SPEC = PartitionSpec.builderFor(SCHEMA)
+      .bucket("data", 16)
+      .build();
+
+  @Before
+  public void writeTestDataFile() throws IOException {
+    File tableDir = temp.newFolder();
+    tableDir.delete();
+    this.table = TestTables.create(tableDir, "test", SCHEMA, SPEC, 2);
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java:49]: I think a better way to break down the class would be to have an abstract `Table createTable(String name, Schema, Spec)` method. Then the `table` and `dataFile` fields don't need to be shared. I also don't think that there is a need to make `records` public either.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java [hunk 7]
```diff
@@ -248,6 +258,26 @@ public void close() throws IOException {
       return iterable;
     }
 
+    @SuppressWarnings("unchecked")
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:261]: Why is this needed?

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java [hunk 8]
```diff
@@ -248,6 +258,26 @@ public void close() throws IOException {
...
+      CloseableIterable<T> iter;
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          // TODO implement value readers for Pig and Hive
+          throw new UnsupportedOperationException("Avro support not yet supported for Pig and Hive");
+        case GENERIC:
+          DeleteFilter deletes = new GenericDeleteFilter(io, currentTask, tableSchema, readSchema);
+          Schema requiredSchema = deletes.requiredSchema();
+          iter = deletes.filter(openTask(currentTask, requiredSchema));
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:272]: Why not return `deletes.filter(...)` here? That would remove the need for `iter` and `break`.

File: mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    int idx = 0;
+
+    for (TestIcebergInputFormats.TestInputFormat.Factory<Record> inputFormat :
+        TestIcebergInputFormats.TESTED_INPUT_FORMATS) {
+      for (String fileFormat : TestIcebergInputFormats.TESTED_FILE_FORMATS) {
+        parameters[idx++] = new Object[] {inputFormat, fileFormat};
+      }
+    }
+
+    return parameters;
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java:74]: Is there a simpler way to configure this? Normally, we build these using literals instead of a block of code.

File: spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java [hunk 10]
```diff
@@ -93,212 +102,4 @@ public void createTable() throws IOException {
...
-    StructLikeSet actual = rowSet(table);
-
-    Assert.assertEquals("Table should contain expected rows", expected, actual);
-  }
-
-  private static StructLikeSet rowSet(Table table) {
-    return rowSet(table, "*");
-  }
-
-  private static StructLikeSet rowSet(Table table, String... columns) {
```
Comment (rdblue) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:271]: This method is what reads the rows from the table using Spark. Deleting this method and using the one in `DeletesReadTest` makes this test suite use the exact same read path as the generics -- `IcebergGenerics`.

You can probably make this method abstract and implement it in both classes to get around this. You'll also need to implement a read using the input format or Hive runner to test the Hive code.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java (hunk 3 above)
Comment (chenjunjiedada) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:136]: The `GenericDeleteFilter` constructor needs `FileIO` as parameter.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java (hunk 7 above)
Comment (chenjunjiedada) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:261]: `deletes.filter(...)` needs this.

File: spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java (hunk 10 above)
Comment (chenjunjiedada) [reply spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:271]: Sorry, I missed this. I just added this back and also use input format to read records.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java (hunk 1 above)
Comment (chenjunjiedada) [reply data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java]: Done.

File: data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java (hunk 4 above)
Comment (chenjunjiedada) [reply data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java]: Updated.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java (hunk 5 above)
Comment (chenjunjiedada) [reply data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java]: Done.

File: data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java (hunk 6 above)
Comment (chenjunjiedada) [reply data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java]: Make sense to me. Updated.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java (hunk 8 above)
Comment (chenjunjiedada) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java]: Updated.

File: mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java (hunk 9 above)
Comment (chenjunjiedada) [reply mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java]: Yes, I just updated it.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java (hunk 2 above)
Comment (chenjunjiedada) [reply data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java]: Removed.

**Review by rdsr — APPROVED** [review]
LGTM!. The changes in MR to incorporate deletes are minimal and clean!

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java [hunk 11]
```diff
@@ -25,49 +25,45 @@
...
+  private List<Record> records;
 
-    table.newAppend()
-        .appendFile(dataFile)
-        .commit();
-  }
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  public abstract Table createTable(String name, Schema schema, PartitionSpec spec) throws IOException;
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java:66]: Why is this `public` and not `protected` to expose it to child classes?

File: data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    return set;
+  }
+
+  @Before
+  public void writeTestDataFile() throws IOException {
+    this.table = createTable("test", SCHEMA, SPEC);
+    generateTestData();
+    table.newAppend()
+        .appendFile(dataFile)
+        .commit();
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java:58]: Why is this `@Before` here and not in the base class?

File: spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java [hunk 13]
```diff
@@ -20,285 +20,111 @@
...
-    DeleteFile eqDeletes = FileHelpers.writeDeleteFile(
-        table, Files.localOutput(temp.newFile()), Row.of(0), dataDeletes, dataSchema);
-
-    List<Pair<CharSequence, Long>> deletes = Lists.newArrayList(
-        Pair.of(dataFile.path(), 3L), // id = 89
-        Pair.of(dataFile.path(), 5L) // id = 121
-    );
+  @Before
+  public void prepareData() throws IOException {
+    this.table = createTable("table", SCHEMA, SPEC);
```
Comment (rdblue) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:90]: Isn't this identical to the one in the generics test? I think this should be in the base class instead.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java (hunk 7 above)
Comment (rdblue) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:261]: Okay, makes sense.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java [hunk 14]
```diff
@@ -248,6 +258,22 @@ public void close() throws IOException {
       return iterable;
     }
 
+    @SuppressWarnings("unchecked")
+    private CloseableIterable<T> open(FileScanTask currentTask, Schema readSchema) {
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          // TODO implement value readers for Pig and Hive
+          throw new UnsupportedOperationException("Avro support not yet supported for Pig and Hive");
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:267]: This error message is for Avro, but I think it should be that Pig and Hive are not supported for any format.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java (hunk 3 above)
Comment (rdblue) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:136]: In that case, this can pass the `FileIO` somehow, or we can work on getting the other PR done before this one. But I don't think we should mix the two features together.

File: mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+  @Override
+  public StructLikeSet rowSet(Table table, String... columns) {
+    Schema projected = table.schema().select(columns);
+    StructLikeSet set = StructLikeSet.create(projected.asStruct());
+    set.addAll(testInputFormat.create(builder.project(projected).conf()).getRecords());
+
+    return set;
+  }
+
+  public TestMrReadDeletes(String inputFormat, FileFormat fileFormat) {
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java:89]: This constructor should be just below the parameters, and then abstract method implementations should come just afterwards.

File: mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        new Object[] { "MapredIcebergInputFormat", FileFormat.PARQUET },
+        new Object[] { "MapredIcebergInputFormat", FileFormat.AVRO },
+        new Object[] { "MapredIcebergInputFormat", FileFormat.ORC },
+    };
+  }
+
+  @Override
+  public Table createTable(String name, Schema schema, PartitionSpec spec) throws IOException {
+    Table table;
+    conf = new Configuration();
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java:66]: Why does this create a new configuration?

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java (hunk 3 above)
Comment (rdblue) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:136]: For now, this could create a new `HadoopFileIO` and use that instead. That would be the easiest path forward.

Comment (rdblue) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:136]: @holdenk just added a `FileIO` instance to this class, so you can use that instead of mixing the two PRs together. Thanks @holdenk!

Comment (chenjunjiedada) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:136]: Thank you @holdenk @rdblue. I updated this.

File: spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java (hunk 13 above)
Comment (chenjunjiedada) [reply spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:90]: Make sense, Updated

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java (hunk 11 above)
Comment (chenjunjiedada) [reply data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java]: Updated to protected.

File: data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java (hunk 12 above)
Comment (chenjunjiedada) [reply data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java]: Updated to the base class.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java (hunk 14 above)
Comment (chenjunjiedada) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java]: This is a copy-paste issue, just updated.

File: mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java (hunk 15 above)
Comment (chenjunjiedada) [reply mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java]: Fixed.

File: mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java (hunk 16 above)
Comment (chenjunjiedada) [reply mr/src/test/java/org/apache/iceberg/mr/TestMrReadDeletes.java]: Hmm, it should be allocated once. I moved allocation to class level.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java [hunk 17]
```diff
@@ -25,53 +25,68 @@
...
-    records.add(record.copy("id", 29, "data", "a"));
-    records.add(record.copy("id", 43, "data", "b"));
-    records.add(record.copy("id", 61, "data", "c"));
-    records.add(record.copy("id", 89, "data", "d"));
-    records.add(record.copy("id", 100, "data", "e"));
-    records.add(record.copy("id", 121, "data", "f"));
-    records.add(record.copy("id", 122, "data", "g"));
+  protected final String testTableName = "test";
+  protected Table testTable;
+  protected DataFile dataFile;
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java:62]: Why are these protected and not private?

My earlier suggestion was to avoid sharing fields with subclasses because the tests are all in this class. Only creating a table and reading it needs to be customized by subclasses, and we can pass in anything necessary to do that.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java [hunk 18]
```diff
@@ -80,21 +95,37 @@ public void testEqualityDeletes() throws IOException {
...
         .commit();
 
     StructLikeSet expected = rowSetWithoutIds(29, 89, 122);
-    StructLikeSet actual = rowSet(table);
+    StructLikeSet actual = rowSet(testTable);
 
     Assert.assertEquals("Table should contain expected rows", expected, actual);
   }
 
+  protected void generateTestData() throws IOException {
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java:110]: I think this method should no longer exist, and there is no reason for it to be accessible to subclasses. It is also not a good idea to write methods like this one that have assumptions about other variables and operate by creating side-effects. It is less of a problem now that it is not called by subclasses. Now, there is just little reason for it to exist when this could be moved into the `@Before` method.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java [hunk 19]
```diff
@@ -25,53 +25,68 @@
...
+  }
+
+  protected abstract Table createTable(String name, Schema schema, PartitionSpec spec) throws IOException;
+
+  protected abstract void dropTable(String name) throws IOException;
+
   @Test
   public void testEqualityDeletes() throws IOException {
-    Schema deleteRowSchema = table.schema().select("data");
+    Schema deleteRowSchema = testTable.schema().select("data");
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java:89]: Renaming `table` -> `testTable` introduces unnecessary changes and doesn't add much value. What was the rationale for doing this?

File: data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TestTables;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.util.StructLikeSet;
+import org.junit.After;
+
+public class GenericReaderDeletesTest extends DeletesReadTest {
+
+  @Override
+  public Table createTable(String name, Schema schema, PartitionSpec spec) throws IOException {
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java:35]: This shouldn't be `public` when the abstract method is `protected`.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java [hunk 21]
```diff
@@ -250,6 +252,22 @@ public void close() throws IOException {
       return iterable;
     }
 
+    @SuppressWarnings("unchecked")
+    private CloseableIterable<T> open(FileScanTask currentTask, Schema readSchema) {
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          // TODO implement value readers for Pig and Hive
+          throw new UnsupportedOperationException("Pig and Hive are not supported for any format");
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:261]: This should be clear that Pig and Hive object models are not supported. Pig and Hive engines can read using generics.

File: mr/src/test/java/org/apache/iceberg/mr/TestIcebergInputFormats.java [hunk 22]
```diff
@@ -370,7 +370,7 @@ public void testCustomCatalog() throws IOException {
     testInputFormat.create(builder.conf()).validate(expectedRecords);
   }
 
-  private abstract static class TestInputFormat<T> {
+  public abstract static class TestInputFormat<T> {
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/TestIcebergInputFormats.java:373]: Why was this changed to `public`?

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java (hunk 17 above)
Comment (chenjunjiedada) [reply data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java]: It should be private now. Updated.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java (hunk 18 above)
Comment (chenjunjiedada) [reply data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java]: Okay, make sense to me. Moved to `@Before`.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java (hunk 19 above)
Comment (chenjunjiedada) [reply data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java]: The `checkstyle` complaints hidden field since the `rowSet` use the parameter name `table`. I just reverted the changes back to change `rowSet` parameter name.

File: data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java (hunk 20 above)
Comment (chenjunjiedada) [reply data/src/test/java/org/apache/iceberg/data/GenericReaderDeletesTest.java]: Fixed.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java (hunk 21 above)
Comment (chenjunjiedada) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:261]: Updated.

File: mr/src/test/java/org/apache/iceberg/mr/TestIcebergInputFormats.java (hunk 22 above)
Comment (chenjunjiedada) [reply mr/src/test/java/org/apache/iceberg/mr/TestIcebergInputFormats.java:373]: Hmm, I thought accessing `TestIcebergInputFormats.TestInputFormat.newFactory` would need it to be public. It turns out not necessary. Updated.

Comment (chenjunjiedada) [reply mr/src/test/java/org/apache/iceberg/mr/TestIcebergInputFormats.java:373]: I tried to minimize the changes to access modifiers in `TestIcebergInputFormat`, the last place that needs the modifier of `TestInputFormat` to be public is `getRecords` method.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java [hunk 23]
```diff
@@ -63,12 +82,20 @@ public void writeTestDataFile() throws IOException {
     records.add(record.copy("id", 122, "data", "g"));
 
     this.dataFile = FileHelpers.writeDataFile(table, Files.localOutput(temp.newFile()), Row.of(0), records);
-
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java:66]: Nit: whitespace-only change.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java [hunk 24]
```diff
@@ -25,31 +25,50 @@
...
-  private DataFile dataFile = null;
+  private DataFile dataFile;
+  private List<Record> records;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
 
   @Before
-  public void writeTestDataFile() throws IOException {
+  public void prepareData() throws IOException {
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java:70]: Nit: unnecessary method renames cause more changed lines than needed.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java [hunk 25]
```diff
@@ -25,31 +25,50 @@
...
+      .bucket("data", 16)
+      .build();
+
+  protected final String testTableName = "test";
+  protected Table table;
 
-  private List<Record> records = null;
-  private DataFile dataFile = null;
+  private DataFile dataFile;
+  private List<Record> records;
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java:64]: Changing the order of these two lines and dropping the default also causes unnecessary changes.

File: data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java [hunk 26]
```diff
@@ -25,31 +25,50 @@
...
+      required(1, "id", Types.IntegerType.get()),
+      required(2, "data", Types.StringType.get())
+  );
+
+  // Partition spec used to create tables
+  public static final PartitionSpec SPEC = PartitionSpec.builderFor(SCHEMA)
+      .bucket("data", 16)
+      .build();
+
+  protected final String testTableName = "test";
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/DeletesReadTest.java:60]: I think this should be private. If it is needed by subclases, it should be passed into methods, not shared. I think this is only used by Spark, so it should be easy to fix.

**Comment by rdblue** [comment]
@chenjunjiedada, I went ahead and fixed the remaining issues and opened a PR against your branch. Could you please take a look?

**Comment by rdblue** [comment]
Merged. Thanks, @chenjunjiedada!
